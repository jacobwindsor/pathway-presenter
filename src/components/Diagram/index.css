.diagram-wrapper {
    height: 900px;
    margin-bottom: 4rem;
}

.diagram-wrapper.isHidden {
    opacity: 0;
}