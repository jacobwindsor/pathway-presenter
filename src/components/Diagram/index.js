import { Pvjs } from '@wikipathways/pvjs';
import React from 'react';
import './index.css';
import PropTypes from 'prop-types';

const Diagram = props => {
  const onPvjsReady = ({ entities, manipulator }) => {
    const { onReady } = props;
    if (onReady) onReady({ entities, manipulator });
  };

  const {
    wpId,
    version,
    showPanZoomControls,
    isHidden,
    slide,
    detailPanelEnabled,
    onEntityClick
  } = props;
  const targets = slide.targets;

  const getEntityIds = entities =>
    entities.map(singleEntity => singleEntity.entityId);

  const highlightedEntities = targets
    .filter(singleTarget => singleTarget.highlighted)
    .map(singleTarget =>
      Object.assign(
        {},
        {
          entityId: singleTarget.entityId,
          color: singleTarget.highlightedColor
        }
      )
    );
  const pannedEntities = getEntityIds(
    targets.filter(singleTarget => singleTarget.panned)
  );
  const zoomedEntities = getEntityIds(
    targets.filter(singleTarget => singleTarget.zoomed)
  );
  const hiddenEntities = getEntityIds(
    targets.filter(singleTarget => singleTarget.hidden)
  );

  return (
    <div className={`diagram-wrapper ${isHidden ? 'isHidden' : ''}`}>
      <Pvjs
        wpId={wpId}
        version={version}
        showPanZoomControls={showPanZoomControls}
        detailPanelEnabled={detailPanelEnabled}
        onEntityClick={onEntityClick}
        highlightedEntities={highlightedEntities}
        pannedEntities={pannedEntities}
        zoomedEntities={zoomedEntities}
        hiddenEntities={hiddenEntities}
        onReady={onPvjsReady}
      />
    </div>
  );
};

Diagram.propTypes = {
  wpId: PropTypes.string.isRequired,
  version: PropTypes.number.isRequired,
  showPanZoomControls: PropTypes.bool.isRequired,
  onReady: PropTypes.func,
  slide: PropTypes.shape({
    targets: PropTypes.array.isRequired
  }).isRequired,
  isHidden: PropTypes.bool,
  detailPanelEnabled: PropTypes.bool,
  onEntityClick: PropTypes.func
};

Diagram.defaultProps = {
  hidden: false,
  detailPanelEnabled: true
};

export default Diagram;
