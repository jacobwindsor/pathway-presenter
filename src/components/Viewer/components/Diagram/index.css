.diagram-wrapper {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    width: 70%;
    height: 70%;
}

.diagram-wrapper.isHidden {
    opacity: 0;
}