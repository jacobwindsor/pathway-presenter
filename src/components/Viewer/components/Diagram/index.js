import { Pvjs } from '@wikipathways/pvjs';
import React, { Component } from 'react';
import './index.css';
import PropTypes from 'prop-types';
import { isMatch, difference } from 'lodash';

class Diagram extends Component {
    constructor(props) {
        super(props);
        this.state = {};
    }

    onPvjsReady = (pvjsRef) => {
        const { onReady } = this.props;
        this.setState({manipulator: pvjsRef.manipulator});
        onReady();
    }

    render() {
        const { wpId, version, showPanZoomControls, isHidden, slide } = this.props;
        const targets = slide.targets;

        const getEntityIds = entities => entities.map(singleEntity => singleEntity.entityId);

        const highlightedEntities = targets
            .filter(singleTarget => singleTarget.highlighted)
            .map(singleTarget => Object.assign({},
                {entityId: singleTarget.entityId, color: singleTarget.highlightedColor})
            );
        const pannedEntities = getEntityIds(targets
            .filter(singleTarget => singleTarget.panned));
        const zoomedEntities = getEntityIds(targets
            .filter(singleTarget => singleTarget.zoomed));
        const hiddenEntities = getEntityIds(targets
            .filter(singleTarget => singleTarget.hidden));

        return (
            <div className={`diagram-wrapper ${isHidden? 'isHidden' : ''}`}>
                <Pvjs about={`http://identifiers.org/wikipathways/WP${wpId}`}
                      version={version}
                      showPanZoomControls={showPanZoomControls}
                      highlightedEntities={highlightedEntities}
                      pannedEntities={pannedEntities}
                      zoomedEntities={zoomedEntities}
                      hiddenEntities={hiddenEntities}
                      onReady={this.onPvjsReady} />
            </div>
        )
    }
}

Diagram.propTypes = {
    wpId: PropTypes.number.isRequired,
    version: PropTypes.number.isRequired,
    showPanZoomControls: PropTypes.bool.isRequired,
    onReady: PropTypes.func,
    slide: PropTypes.object.isRequired,
    isHidden: PropTypes.bool
};

Diagram.defaultProps = {
    hidden: false
};

export default Diagram;