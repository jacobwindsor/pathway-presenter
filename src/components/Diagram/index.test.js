import React from 'react';
import { shallow } from 'enzyme';
import Diagram from './index';
import toJson from 'enzyme-to-json';

describe('The diagram component. Renders just the Pvjs diagram', () => {
  it('renders without crashing', () => {
    const comp = shallow(
      <Diagram
        wpId={'WP4'}
        version={0}
        showPanZoomControls={true}
        slide={{ targets: [] }}
      />
    );

    expect(toJson(comp)).toMatchSnapshot();
  });

  it('should have the diagram-wrapper class', () => {
    const comp = shallow(
      <Diagram
        wpId="WP4"
        version={0}
        showPanZoomControls
        slide={{ targets: [] }}
      />
    );
    expect(comp.find('.diagram-wrapper')).toHaveLength(1);
  });

  it('should have only one child and that is Pvjs', () => {
    const comp = shallow(
      <Diagram
        wpId="WP4"
        version={0}
        showPanZoomControls
        slide={{ targets: [] }}
      />
    );
    expect(comp.children()).toHaveLength(1);
  });

  it('should pass in action (manipulation) props correctly to Pvjs', () => {
    const slide = {
      targets: [
        {
          highlighted: true,
          highlightedColor: 'red',
          entityId: '1234',
          panned: true
        },
        {
          entityId: '5678',
          highlighted: false,
          hidden: true
        },
        {
          entityId: '9123',
          highlighted: true,
          highlightedColor: 'blue',
          panned: true,
          zoomed: true
        },
        {
          entityId: '9234',
          highlighted: false,
          panned: true,
          zoomed: true
        }
      ]
    };

    const comp = shallow(
      <Diagram wpId="WP4" version={0} showPanZoomControls slide={slide} />
    );

    const shouldBeHighlighted = [
      {
        entityId: '1234',
        color: 'red'
      },
      {
        entityId: '9123',
        color: 'blue'
      }
    ];

    const shouldBeZoomed = ['9123', '9234'];
    const shouldBePanned = ['1234', '9123', '9234'];
    const shouldBeHidden = ['5678'];

    const pvjs = comp.children().first();
    expect(pvjs.prop('highlightedEntities')).toEqual(shouldBeHighlighted);
    expect(pvjs.prop('zoomedEntities')).toEqual(shouldBeZoomed);
    expect(pvjs.prop('pannedEntities')).toEqual(shouldBePanned);
    expect(pvjs.prop('hiddenEntities')).toEqual(shouldBeHidden);
  });
});
