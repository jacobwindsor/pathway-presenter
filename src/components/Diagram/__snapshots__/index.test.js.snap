// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The diagram component. Renders just the Pvjs diagram renders without crashing 1`] = `
<div
  className="diagram-wrapper "
>
  <Pvjs
    detailPanelEnabled={true}
    hiddenEntities={Array []}
    highlightedEntities={Array []}
    onReady={[Function]}
    pannedEntities={Array []}
    showPanZoomControls={true}
    version={0}
    wpId="WP4"
    zoomedEntities={Array []}
  />
</div>
`;
